### PR TITLE
Batch mode

### PR DESCRIPTION
**Pull request by fayway**
Hi @manekinekko,

I tried to continue on your wip, I mainly used RxJS operators without making too much changes in `BluetoothCore` to ensure a maximum backward compatibility.

I also adapted the examples of the starter to be able to test, but I think it's better to present in the starter two distinct examples: A simple one with only one service/characteristic and an advanced example with a device that supports several services. This way, examples will be adapted to the needs (from minimalist to more complicated).

By the way, I reused the fake test objects to simulate a super device that supports all the services of the wip dashboard, to activate it, just use `fakeBleCore` as the factory function in [app.module.ts](https://github.com/fayway/angular-web-bluetooth/blob/batch-mode/src/app/app.module.ts#L39)

```ts
const PROVIDERS = [
  {
    provide: BluetoothCore,
    useFactory: fakeBleCore, // bleCore or fakeBleCore
    deps: [BrowserWebBluetooth, ConsoleLoggerService]
  },
  DashboardService
];
```

localhost_4200_

Now I wonder if we should push the dashboard example further, like giving the possibility to switch between a global dashboard device and a local component device, handle more "unsupported service/characteristic" scenarios and that kind of stuff...

How do you see things from your side?

### CODE REVIEW

**Comment by manekinekko** [comment]
This is amazing work!!! I will schedule some time to check the PR. Thank so much for looking at this highly requested feature 🙏

**Review by manekinekko — APPROVED** [review]

**Comment by manekinekko** [comment]
@fayway would it be possible to update the readme file to add instructions about how to use the batch mode?

**Comment by fayway** [comment]
Well, actually I'm not very satisfied with this PR 😅  I push it despite of that so I can have your feedbacks.
They are some parts we can improve by moving them from DashboardService to BluetoothCore instead so clients will not have to duplicate the same streams manipulations. 

When studying the Bluetooth Web API, the only method I found that takes multiple uuids and return multiple values is [BluetoothRemoteGATTService.getCharacteristics](https://developer.mozilla.org/en-US/docs/Web/API/BluetoothRemoteGATTService/getCharacteristics).

There is also [BluetoothRemoteGATTServer.getPrimaryServices](https://developer.mozilla.org/en-US/docs/Web/API/BluetoothRemoteGATTServer/getPrimaryServices) that takes a single uuid but can return multiple services! Personally I don't understand how it's possible. (Don't forget I'm just simulating all this stuff 😁)

Anyway, if we just use `BluetoothRemoteGATTServer.getPrimaryService` (without `s`) and `BluetoothRemoteGATTService.getCharacteristics` (with `s`), what are the new methods we can introduce in `BluetoothCore` (bluetooth.service.ts) ?

For example, what do you think of, in plus of the regular `value$(options: ReadValueOptions)`, also add a `values$(options: BatchReadOptions)` that looks like that:

```ts
type BatchReadOptions = {
  acceptAllDevices?: boolean;
  services: [{
    serviceUUID: string,
    characteristicUUIDs: string[]
  }]
};

type DetailedValue = {
  service: BluetoothServiceUUID,
  characteristic: BluetoothCharacteristicUUID,
  value: DataView
};

export class BluetoothCore  {
    values$(options: BatchReadOptions): Observable<DetailedValue> {
       // Adapt BatchReadOptions to Ble Web API and do the RxJS magic
    }
}
```

and of course add the plural version for all other methods (`getPrimaryServices`,`getCharacteristics`...)

**Comment by manekinekko** [comment]
That's awesome! If you have time to make the API better, please do. I know this requires a lot of time and efforts to do and I didn't wanted to add more work to your plate.

As per the batch API, your suggestion LGTM. That's what folks have been requesting. 

One question though, how errors are going to be handled? Say, we request read from 5 characteristics and 3 of them return different errors (1 error for each characteristic). We should be able to identify what characteristic (or service) triggered which error so we can reflect that on the UI.

**Comment by fayway** [comment]
For errors, I just catch them and push a [SnackBar notification](
https://github.com/manekinekko/angular-web-bluetooth/pull/76/files#diff-2644ac105391147abec80a8eca9f3178a830c835111ac6c361006bf26d913638R90)

```ts
this.ble.getGATT$()
    .pipe(
    mergeMap((gatt: BluetoothRemoteGATTServer) => this.ble.getPrimaryService$(gatt, c.service)
        .pipe(catchError(error => {
        this.hasError(error);
        return EMPTY;
        }))),
    mergeMap((gattService: BluetoothRemoteGATTService) => this.ble.getCharacteristic$(gattService, c.characteristic)
        .pipe(catchError(error => {
        this.hasError(error);
        return EMPTY;
    }))),
    ....
```
This is what it gives using the Android simulator that only supports `battery_service`

Screenshot 2021-03-22 at 17 37 13

Do you prefer a more visual indicator ?

For merging the PR as it is in a first time and make the core update at a second time,  I think that we should at least have a minimalist example without the batch mode. 

So I suggest to have in the started 2 main cards:

1. First card with the battery service and its own connect/disconnect (like before, this battery service is the simplest one we can simulate using the Android simulator)
2. A second card with 3 sub cards: the Step Counter, temperature and humidity with a single connect/disconnect to demonstrates the batch mode.

And since it's very uncommon to have a Step counter service in a weather station, I will not implement it in my fake generator to show how we handle errors.

For me personally, examples are the key...
Is it ok like this?

**Comment by manekinekko** [comment]
> For errors, I just catch them and push a SnackBar notification

I think we could probably simply errors handling. That's a lot of boilerplate the write and I'm afraid folks who are not familiar with Observables or RxJs will be lost. Can we abstract this code into a helper function?

> Is it ok like this?

Yes, I do agree that having examples will help a lot.

**Comment by fayway** [comment]
Hi @manekinekko,

Some new commits with:

1 - A simple BLE service example like it was before involving only the Battery Level Service 

simple-mode

2- And a batch mode demo in a dedicated route (with its own components/services)

batch-mode

I'm not sure what generic helper function we can have for errors handling, in this new commit, I suggest to stream errors to sub components to let them do whatever they want to do with... like in this example, to display a visual indicator when the service is not supported:

unsupported-service

What do you think?

File: README.md [hunk 1]
```diff
@@ -153,31 +152,100 @@ export class BatteryLevelService {
 }
 ```
 
+## 2.c) subscribe to multiple services/characteristics on the same device 
+
+While waiting for a more native implementation, you can perform the batch operations mode my calling `discover$()` only once and then merge all streamed services/characteristics using RxJS as a workaround:
```
Comment (manekinekko) [inline README.md:157]:
```suggestion
You can perform the batch operations mode my calling `discover$()` only once and then merge all streamed services/characteristics using RxJS:
```

File: README.md [hunk 2]
```diff
@@ -153,31 +152,100 @@ export class BatteryLevelService {
...
+import { BluetoothCore } from '@manekinekko/angular-web-bluetooth';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class BleBatchService {
+  private config;
+
+  constructor(public ble: BluetoothCore) {
+    // 1) define supported services/characteristics that interest you
```
Comment (manekinekko) [inline README.md:172]:
```suggestion
    // 1) define services/characteristics that you want to read data from 
```

File: README.md [hunk 3]
```diff
@@ -153,31 +152,100 @@ export class BatteryLevelService {
...
+  }
+  
+  values() {
+    // 2) call discover$ to trigger the discovery process
+    this.ble.discover$({
+      acceptAllDevices: true,
+      optionalServices: config.map(c => c.service)
+    }).subscribe(...);
+  
+    // 3) for every service/characteristic couple
```
Comment (manekinekko) [inline README.md:191]:
```suggestion
    // 3) for every service or characteristic pair
```

File: README.md [hunk 4]
```diff
@@ -153,31 +152,100 @@ export class BatteryLevelService {
...
 ## Blog post
 
-Checkout my post on medium.
-
-<p align="center">
-  <a href="https://dev.to/angular/the-web-bluetooth-module-for-angular-314b">
-    <img src="https://cloud.githubusercontent.com/assets/1699357/21696708/7e33cca4-d38f-11e6-8a03-6833b88e82fa.png" >
-  </a>
-</p>
+Checkout my full [blog post on dev.to](https://dev.to/angular/the-web-bluetooth-module-for-angular-314b) about how to use this package in your app.
```
Comment (manekinekko) [inline README.md:244]:
```suggestion
Checkout the full [blog post on dev.to](https://dev.to/angular/the-web-bluetooth-module-for-angular-314b) about how to use this package in your app.
```

File: README.md [hunk 5]
```diff
@@ -153,31 +152,100 @@ export class BatteryLevelService {
...
 
 ## Have a PR?
 
-All contributions are welcome ;)
+All contributions are welcome. Here are few [open issues](https://github.com/manekinekko/angular-web-bluetooth/issues?q=is%3Aissue+is%3Aopen+label%3A%22help+wanted%22) that I need help with ;)
```
Comment (manekinekko) [inline README.md:248]:
```suggestion
All contributions are welcome. Here are few [open issues](https://github.com/manekinekko/angular-web-bluetooth/issues?q=is%3Aissue+is%3Aopen+label%3A%22help+wanted%22) that we need help with ;)
```

**Review by manekinekko — CHANGES_REQUESTED** [review]

**Pull request merged**